packages/decimal/src/decimal256.rs

```diff
@@ -0,0 +1,1242 @@
+
+            while n > 1 {
+                if n % 2 == 0 {
+                    x = x.clone().mul(env, &x);
+                    n /= 2;
+                } else {
+                    y = x.clone().mul(env, &y);
+                    x = x.clone().mul(env, &x);
+                    n = (n - 1) / 2;
+                }
+            }
+
+            x.mul(env, &y)
+        }
+
+        inner(env, self, exp)
+    }
+
+    pub fn inv(&self, env: &Env) -> Option<Self> {
+        if self.is_zero(env) {
+            None
+        } else {
+            let fractional_squared =
+                U256::from_u128(env, 1_000_000_000_000_000_000_000_000_000_000_000_000);
+            Some(Decimal256(fractional_squared.div(&self.0)))
+        }
+    }
+
+    pub fn from_ratio(env: &Env, numerator: impl Into<U256>, denominator: impl Into<U256>) -> Self {
+        match Decimal256::checked_from_ratio(env, numerator, denominator) {
+            Ok(ratio) => ratio,
+            Err(Error::DivideByZero) => panic!("Denominator must not be zero"),
+        }
+    }
+
+    pub fn to_u128_with_precision(&self, precision: impl Into<i32>) -> u128 {
+        let value = self.atomics().unwrap();
+        let precision = precision.into();
+
+        let divisor = 10u128.pow((self.decimal_places() - precision) as u32);
+        value / divisor
+    }
+
+    fn multiply_ratio(
+        &self,
+        env: &Env,
+        numerator: Decimal256,
+        denominator: Decimal256,
+    ) -> Decimal256 {
+        Decimal256::from_ratio(env, self.0.mul(&numerator.0), denominator.0)
+    }
+
+    fn checked_from_ratio(
+        env: &Env,
+        numerator: impl Into<U256>,
+        denominator: impl Into<U256>,
+    ) -> Result<Self, Error> {
+        let numerator = numerator.into();
+        let denominator = denominator.into();
+
+        if denominator == U256::from_u128(env, 0) {
+            return Err(Error::DivideByZero);
+        }
+
+        let ratio = numerator
+            .mul(&U256::from_u128(env, 1_000_000_000_000_000_000))
+            .div(&denominator);
+        Ok(Decimal256(ratio))
+    }
+
+    pub fn abs_diff(self, env: &Env, other: Self) -> Self {
+        let diff = self
+            .0
+            .to_u128()
+            .unwrap()
+            .abs_diff(other.0.to_u128().unwrap());
+        Self(U256::from_u128(env, diff))
+    }
+
+    pub fn div_by_u256(&self, rhs: U256) -> Self {
+        Decimal256(self.0.div(&rhs))
+    }
+}
+
+impl Add for Decimal256 {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Decimal256(self.0.add(&other.0))
+    }
+}
+
+impl Sub for Decimal256 {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        Decimal256(self.0.sub(&other.0))
+    }
+}
+
+impl Decimal256 {
+    pub fn mul(&self, env: &Env, other: &Self) -> Self {
+        let result = self
+            .numerator()
+            .mul(&other.numerator())
+            .div(&other.denominator(env));
```

Comment:
One function call less.

```suggestion
            .div(&U256::from_u128(env, 1_000_000_000_000_000_000));
```